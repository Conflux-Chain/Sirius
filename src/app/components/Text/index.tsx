import React from 'react';
import PropTypes from 'prop-types';
import { Tooltip, Text } from '@cfxjs/react-ui';
import { TooltipProps } from '@cfxjs/react-ui/dist/tooltip/tooltip';
import { TextProps as ReactUITextProps } from '@cfxjs/react-ui/dist/text/text';
import styled from 'styled-components/macro';
import { selectText } from './../../../utils/util';

type TextProps = {
  children?: React.ReactNode;
  maxwidth?: string;
  maxCount?: number;
  hoverValue?: React.ReactNode;
  hovercolor?: string;
  tooltip?: Partial<TooltipProps>;
} & Partial<ReactUITextProps>;
type NativeAttrs = Omit<React.HTMLAttributes<any>, keyof TextProps>;
export declare type Props = TextProps & NativeAttrs;

const Wrapper = styled(Text)<any>`
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  vertical-align: bottom;
  max-width: ${props => props.maxwidth};
  display: ${props =>
    props.maxwidth === undefined ? 'inherit' : 'inline-block'};
  cursor: pointer;
  font-weight: 400;
  &:hover {
    font-weight: 500;
    color: ${props => props.hovercolor};
  }
` as React.FC<Props>;

// note:
// 1. maxwidth priority is higher than maxCount
// 2. maxCount only apply to string
// 3. if hoverValue is provided, use hoverValue as Tooltip text, otherwise use children
//    if text of prop tooltip is provided, use as Tooltip text
const TextComponent = ({
  children,
  maxwidth,
  maxCount,
  hoverValue,
  tooltip,
  ...props
}: Props) => {
  let child: React.ReactNode = children;
  if (maxwidth === undefined && maxCount && typeof children === 'string') {
    child = String.prototype.substr.call(children, 0, maxCount) + '...';
  }
  const tooltipText = (
    <div onClick={e => selectText(e.currentTarget)}>
      {hoverValue || children}
    </div>
  );
  return (
    <Tooltip
      portalClassName="siriui-tooltip-square"
      text={tooltipText}
      {...tooltip}
    >
      <Wrapper maxwidth={maxwidth} {...props}>
        {child}
      </Wrapper>
    </Tooltip>
  );
};

TextComponent.defaultProps = {
  hovercolor: '#0054fe',
};

TextComponent.propTypes = {
  maxCount: PropTypes.number,
  maxwidth: PropTypes.string,
  hoverValue: PropTypes.node,
  hovercolor: PropTypes.string,
  tooltip: PropTypes.object,
};

export default TextComponent;
