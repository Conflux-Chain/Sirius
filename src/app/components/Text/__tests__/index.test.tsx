import React from 'react';
import { render } from '@testing-library/react';
import Text from '..';

describe('<Text  />', () => {
  it('should match snapshot', () => {
    const loadingIndicator = render(
      <Text maxWidth="40px">
        <span>1234567890</span>
      </Text>,
    );
    expect(loadingIndicator.container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when set maxCount', () => {
    const loadingIndicator = render(<Text maxCount={5}>1234567890</Text>);
    expect(loadingIndicator.container.firstChild).toMatchSnapshot();
  });
});
