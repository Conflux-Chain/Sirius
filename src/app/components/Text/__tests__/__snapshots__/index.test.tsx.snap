// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text  /> should match snapshot 1`] = `
.c0 .tooltip-content.sirius-tooltip {
  border-radius: 0;
  padding: 0.2857rem 0.7143rem;
  font-weight: 400;
  border-radius: 0;
  font-size: 12px;
  color: #cdcdcd;
  line-height: 1.1429rem;
}

.c1 .text.sirius-text {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  vertical-align: bottom;
  max-width: 40px;
  display: inline-block;
  cursor: pointer;
}

<span
  class="c0"
>
  <div
    class="jsx-726172838 tooltip-wrapper"
  >
    <div
      class="jsx-726172838 tooltip "
    >
      <span
        class="c1"
      >
        <p
          class="jsx-3390086683  text sirius-text"
        >
          <span>
            1234567890
          </span>
        </p>
      </span>
    </div>
  </div>
</span>
`;

exports[`<Text  /> should match snapshot when set maxCount 1`] = `
.c0 .tooltip-content.sirius-tooltip {
  border-radius: 0;
  padding: 0.2857rem 0.7143rem;
  font-weight: 400;
  border-radius: 0;
  font-size: 12px;
  color: #cdcdcd;
  line-height: 1.1429rem;
}

.c1 .text.sirius-text {
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  vertical-align: bottom;
  display: inherit;
  cursor: pointer;
}

<span
  class="c0"
>
  <div
    class="jsx-726172838 tooltip-wrapper"
  >
    <div
      class="jsx-726172838 tooltip "
    >
      <span
        class="c1"
      >
        <p
          class="jsx-3390086683  text sirius-text"
        >
          12345...
        </p>
      </span>
    </div>
  </div>
</span>
`;
